### PR TITLE
ARROW-5311: [C++] use more specific error status types in take

### PR DESCRIPTION
**Pull request by bkietz**
Now emits

- TypeError when indices are not integral
- KeyError when an index is out of bounds

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
@bkietz you will probably need to update the python test to change ValueError in TypeError here: https://github.com/apache/arrow/blob/master/python/pyarrow/tests/test_compute.py#L93 (and the ValueError to KeyError in the two out of bounds cases a bit higher in the file)

**Comment by bkietz** [comment]
@jorisvandenbossche thanks, good catch

**Review by pitrou — CHANGES_REQUESTED** [review]
I've rather have a dedicated `IndexError` for out-of-bounds issues.

File: cpp/src/arrow/status.h [hunk 1]
```diff
@@ -193,6 +194,13 @@ class ARROW_EXPORT Status {
     return Status(StatusCode::Invalid, util::StringBuilder(std::forward<Args>(args)...));
   }
 
+  /// Return an error status when a container's capacity would exceed its limits
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/status.h]: I suppose this is a left-over from copy paste? (from CapacityError)

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for the update. There is a remaining thing to do: in `ConvertPyError()` (`src/arrow/python/common.cc`), return `StatusCode::IndexError` if the Python exception matches `PyExc_IndexError`.

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
The Java Gandiva failure is unrelated, will merge.

File: c_glib/arrow-glib/error.cpp [hunk 2]
```diff
@@ -37,78 +37,68 @@ G_BEGIN_DECLS
  * values.
  */
 
-G_DEFINE_QUARK(garrow-error-quark, garrow_error)
+G_DEFINE_QUARK(garrow - error - quark, garrow_error)
```
Comment (kou) [inline c_glib/arrow-glib/error.cpp]: This should not be included spaces around `-`.
`G_DEFINE_QUARK` is a macro. `garrow-error-quark` is used as an identifier.

Arrow GLib doesn't use the same coding style as Arrow C++. So we don't need to use `clang-format`.
I'll send a pull request to revert this change.

**Pull request closed without merging**